### PR TITLE
Migrate deprecated `failure-domain.beta.kubernetes.io` labels to `topology.kubernetes.io`

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area usability
/kind bug

**What this PR does / why we need it**:
Starting from Kubernetes 1.28, the old deprecated failure-domain labels are no longer present. However, for clusters with very old volumes, they might still be part of the `nodeSelectorTerms`

**Special notes for your reviewer**:
/cc @timuthy @kon-angelo @adenitiu 

**Release note**:

```bugfix operator
A bug has been fixed which prevented pods from starting on clusters of at least `1.28` if they were using old `PersistentVolume`s created with the deprecated `failure-domain.beta.kubernetes.io/{zone,region}` labels.
```

### CODE REVIEW

File: cmd/gardenlet/app/app.go [hunk 1]
```diff
@@ -760,6 +765,46 @@ func reconcileLabelsForPVCMigrations(ctx context.Context, log logr.Logger, seedC
...
+			patch := client.MergeFrom(persistentVolume.DeepCopy())
+
+			delete(persistentVolume.Labels, corev1.LabelFailureDomainBetaRegion)
+			delete(persistentVolume.Labels, corev1.LabelFailureDomainBetaZone)
+
+			if persistentVolume.Spec.NodeAffinity != nil && persistentVolume.Spec.NodeAffinity.Required != nil {
+				for i, term := range persistentVolume.Spec.NodeAffinity.Required.NodeSelectorTerms {
+					for j, expression := range term.MatchExpressions {
+						if expression.Key == corev1.LabelFailureDomainBetaRegion {
+							persistentVolume.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions[j].Key = corev1.LabelTopologyRegion
```
Comment (vpnachev) [inline cmd/gardenlet/app/app.go]: PV node affinities were immutable until k8s 1.27, I think this should be executed only if the seed clusters is of k8s version 1.27 or higher. 

ref: https://github.com/kubernetes/kubernetes/pull/115391

File: cmd/gardenlet/app/app.go [hunk 2]
```diff
@@ -760,6 +765,46 @@ func reconcileLabelsForPVCMigrations(ctx context.Context, log logr.Logger, seedC
...
+						}
+
+						if expression.Key == corev1.LabelFailureDomainBetaZone {
+							persistentVolume.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions[j].Key = corev1.LabelTopologyZone
+						}
+					}
+				}
+			}
+
+			return seedClient.Patch(ctx, &persistentVolume, patch)
```
Comment (timuthy) [inline cmd/gardenlet/app/app.go]: Can you add a check if changes are really necessary to prevent sending empty patches?

**Review by timuthy — CHANGES_REQUESTED** [review]

File: cmd/gardenlet/app/app.go (hunk 1 above)
Comment (rfranzke) [reply cmd/gardenlet/app/app.go]: That's a nice catch, thank you!

**Review by timuthy — COMMENTED** [review]
Thank you! 👏 
/lgtm

**Review by kon-angelo — COMMENTED** [review]
/lgtm

**Pull request merged**